File: count-the-number-of-complete-components.java
class Solution {
    public int countCompleteComponents(int n, int[][] edges) {
        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();
        for(int i=0;i<n;i++){
            adj.add(new ArrayList<>());
        }
        for(int []edge : edges){
            int a = edge[0];
            int b= edge[1];
            adj.get(a).add(b);
            adj.get(b).add(a); 
        }
        int []vis= new int[n];
        int count=0;
        for(int i=0;i<n;i++){
            if(vis[i]==0){
                List<Integer> ls = new ArrayList<>();
                int edgecount = dfs(i,vis,adj,ls);
                int nodecount = ls.size();
                if(nodecount*(nodecount-1) ==  edgecount)count++;
            }
        }
        return count;
    }
    public int  dfs(int node,int []vis,ArrayList<ArrayList<Integer>> adj,List<Integer> ls){
        vis[node]=1;
        ls.add(node);
        int ct=0;
        for(int i :adj.get(node)){
            if(vis[i]==0){
                ct+=dfs(i,vis,adj,ls);
            }
            ct++;
        }
        return ct;
    }
}

File: count-the-number-of-complete-components.py
class Solution:
    def countCompleteComponents(self, n: int, edges: List[List[int]]) -> int:
        graph = [[] for _ in range(n)]
        
        for u, v in edges:
            graph[u].append(v)
            graph[v].append(u)

        visited = [False] * n
        ans = 0

        def dfs(node):
            visited[node] = True
            vertices.append(node)

            for nei in graph[node]:
                if not visited[nei]:
                    dfs(nei)

        for i in range(n):
            if not visited[i]:
                vertices = []
                dfs(i)

                k = len(vertices)
                edge_count = sum(len(graph[v]) for v in vertices) // 2

                if edge_count == k * (k - 1) // 2:
                    ans += 1

        return ans
